operator/pkg/util/cert/util.go

```diff
@@ -0,0 +1,246 @@
+	}
+	err = createNamespace(client, ns, dryRun)
+	if err != nil {
+		return err
+	}
+	err = createSecret(client, ns, rootCertBytes, intermidiateKeyBytes, intermidiateCertBytes, intermidiateCertBytes, dryRun)
+	if err != nil {
+		return err
+	}
+	opts.Org = org + ".intermediate1"
+	intermidiateCertBytes1, intermidiateKeyBytes1, ckErr := util.GenCertKeyFromOptions(opts)
+	if ckErr != nil {
+		return ckErr
+	}
+	opts.Org = org + ".intermediate2"
+	intermidiateCertBytes2, intermidiateKeyBytes2, ckErr := util.GenCertKeyFromOptions(opts)
+	if ckErr != nil {
+		return ckErr
+	}
+	err = saveCreds(outputDir, rootCertBytes, intermidiateCertBytes1, intermidiateKeyBytes1, intermidiateCertBytes2, intermidiateKeyBytes2)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func saveCreds(certDir string, rootCerts []byte, certPem1 []byte, privPem1 []byte, certPem2 []byte, privPem2 []byte) error {
+	err := ioutil.WriteFile(certDir+"/"+"root-cert.pem", rootCerts, 0644)
+	if err != nil {
+		return (fmt.Errorf("could not write output certificate: %s", err))
+	}
+
+	err = ioutil.WriteFile(certDir+"/"+"ca-cert-1.pem", certPem1, 0644)
+	if err != nil {
+		return (fmt.Errorf("could not write output certificate: %s", err))
+	}
+
+	err = ioutil.WriteFile(certDir+"/"+"ca-key-1.pem", privPem1, 0600)
+	if err != nil {
+		return (fmt.Errorf("could not write output private key: %s", err))
+
+	}
+	err = ioutil.WriteFile(certDir+"/"+"ca-cert-2.pem", certPem2, 0644)
+	if err != nil {
+		return (fmt.Errorf("could not write output certificate: %s", err))
+	}
+
+	err = ioutil.WriteFile(certDir+"/"+"ca-key-2.pem", privPem2, 0600)
+	if err != nil {
+		return (fmt.Errorf("could not write output private key: %s", err))
+
+	}
+	log.Infof("root cert, intermediate ca cert and key have been generated and saved in %v", certDir)
+	return nil
+}
+
+// WriteCert writes the pem-encoded certificate data to certPath.
+// The certificate file will be created with file mode 0644.
+// If the certificate file already exists, it will be overwritten.
+// The parent directory of the certPath will be created as needed with file mode 0755.
+func WriteCert(certPath string, data []byte) error {
+	if err := os.MkdirAll(filepath.Dir(certPath), os.FileMode(0755)); err != nil {
+		return err
+	}
+	return ioutil.WriteFile(certPath, data, os.FileMode(0644))
+}
+
+// CreateTempDirForCert is a function that creates a temporary directory under /etc/istio/tmp
+func CreateTempDirForCert(istioDir string, dirName string) (string, error) {
+	tempDir := path.Join(istioDir, "tmp")
+	// creates target folder if not already exists
+	if err := os.MkdirAll(tempDir, 0700); err != nil {
+		return "", fmt.Errorf("failed to create directory %q", tempDir)
+	}
+
+	tempDir, err := ioutil.TempDir(tempDir, dirName)
+	if err != nil {
+		return "", fmt.Errorf("couldn't create a temporary directory due to %s", err)
+	}
+	return tempDir, nil
+}
+
+func createNamespace(client kubernetes.Interface, ns string, dryRun bool) error {
```

Comment:
I hope operator has some method already to create a namespace or check if it exists before. 

The first stage of installing Istio is applying the 'base' - which creates the namespace, and may customize various things ( labels, etc). Certs should be created after 'base' is applied, and probably fail of the namespace is missing.

Reply:
Moved to after namespace is created